doc/vidb.8
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "VIDB 8"
.TH VIDB 8 "2004-10-16" "SFS 0.8pre" "SFS 0.8pre"
.SH "NAME"
vidb \- manually edit SFS user\-authentication database file
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
vidb [\-w] [\-R] {\-S | \-a [\-f \fIfile\fR] | [\-e \fIeditor\fR]} \fIsfs-users-file\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBvidb\fR manually edits an \s-1SFS\s0 user-authentication file
, acquiring locks to prevent concurrent updates from
overwriting each other.  If \fBsfsauthd\fR has been compiled with
Sleepycat database (\fIhttp://www.sleepycat.com/\fR) support, and the
name of the file ends in \fI.db/\fR, \fBvidb\fR will consider the
user authentication file to be a database directory, and convert the
contents into regular \s-1ASCII\s0 text for editing.  If the name of the file
ends in \fI.db\fR, \fIvidb\fR assumes the user authentication file
is database file (unless the pathname corresponds to an existing
directory).  Note that database files (as opposed to directories) are
required to be read\-only, and thus cannot be updated by vidb.
.SH "OPTIONS"
.IX Header "OPTIONS"
\&\fBvidb\fR has the following options:
.IP "\-a [\-f \fIfile\fR]" 4
.IX Item "-a [-f file]"
The \fB\-a\fR option \fIadds\fR \s-1SFS\s0 user records in text form to a
database.  The records are taken from standard input, or from
\&\fIfile\fR if specified.  Records for an existing user or group will
replace the values already in the database.  Unlike \fBvidb\fR's
ordinary mode of operation, \fB\-a\fR does not add all records
atomically.  In the event of a system crash, some but not all of the
records may have been added to the database.  Simply re-running the
same \fBvidb\fR command after a crash is perfectly safe, however,
since previously added entries will just be overwritten (by
themselves) the second time through.  For database files, because
\&\fB\-a\fR does not accumulate records into one large transaction, it
can be significantly more efficient than simply adding the records in
an editor, using \fBvidb\fR's ordinary mode of operation.
.IP "\-e \fIeditor\fR" 4
.IX Item "-e editor"
Specifies the editor to use for editing the file.  The default is to
use the command specified by the environment variable \fB\s-1EDITOR\s0\fR.
If there is no environment variable and \fB\-e\fR is not specified,
\&\fBvidb\fR uses \fBvi\fR.
.IP "\-w" 4
.IX Item "-w"
One of the points of \fBvidb\fR is to avoid concurrent edits to
the database and the corresponding inconsistencies that might result.
Ordinarily, if the database is already being edited, \fBvidb\fR
will just exit with an error message.  The \fB\-w\fR flag tells
\&\fBvidb\fR to wait until it can acquire the lock on the database
before launching the editor.
.IP "\-R" 4
.IX Item "-R"
Runs \fIcatastrophic recovery\fR on the database environment.  (For
those familiar with Sleepycat database software, this corresponds to
the \fB\-c\fR flag of the \fBdb_recover\fR utility, or the
\&\fB\s-1DB_RECOVER_FATAL\s0\fR flag of the \s-1API\s0.)  Essentially, \fB\-R\fR
replays all of the database log records present in the supporting
files directory.  You may need to use this, for example, when
restoring a database from backup tapes if the log files were backed up
more recently than the entire database.  The \fB\-R\fR has no effect
on flat text databases, or if the \fB\-S\fR has been specified.
\&\fBWarning:\fR  The authors have encountered bugs in the
catastrophic recovery code of at least some versions of the Sleepycat
database package.  As a precaution, before attempting to use
\&\fB\-R\fR, we strongly recommend salvaging whatever records possible
from the database file itself using \fBvidb \-S\fR
\&\fIsfs-users-file\fR>\fBsaved_sfs_users\fR.  If, subsequently,
the \fB\-R\fR option corrupts the database, you can at least salvage
some of the records from the \fIsaved_sfs_users\fR file.
.IP "\-S" 4
.IX Item "-S"
Attempt to \fIsalvage\fR a database file with corrupt or lost logs by
dumping the contents of the database itself.  Ordinarily, databases
use write-ahead logging.  Before opening a database file, both
\&\fBvidb\fR and \fBsfsauthd\fR attempt to recover from any
previous incomplete transactions using the log.  The \fB\-S\fR
option opens and prints out the contents of a database without regard
to the log files.  This is useful if you have lost the log files or
are worried that they are corrupt, or if you wish to examine the
contents of a database you have read but not write permission to.
Ordinarily, however, if you wish to dump the contents of a database to
standard output, use the command \fBvidb \-e cat\fR
\&\fIsfs-users-file\fR.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIdirsearch\fR\|(1), \fInewaid\fR\|(1), \fIrex\fR\|(1), \fIsfsagent\fR\|(1), \fIsfskey\fR\|(1), \fIssu\fR\|(1), \fIsfs_config\fR\|(5), \fIsfs_hosts\fR\|(5), \fIsfs_srp_params\fR\|(5), \fIsfs_users\fR\|(5), \fIsfsauthd_config\fR\|(5), \fIsfscd_config\fR\|(5), \fIsfsrosd_config\fR\|(5), \fIsfsrwsd_config\fR\|(5), \fIsfssd_config\fR\|(5), \fIsfs_environ\fR\|(7), \fIfunmount\fR\|(8), \fInfsmounter\fR\|(8), \fIsfsauthd\fR\|(8), \fIsfscd\fR\|(8), \fIsfsrosd\fR\|(8), \fIsfsrwcd\fR\|(8), \fIsfsrwsd\fR\|(8), \fIsfssd\fR\|(8)
.PP
The full documentation for \fB\s-1SFS\s0\fR is maintained as a Texinfo
manual.  If the \fBinfo\fR and \fB\s-1SFS\s0\fR programs are properly installed
at your site, the command \fBinfo \s-1SFS\s0\fR
should give you access to the complete manual.
.PP
For updates, documentation, and software distribution, please
see the \fB\s-1SFS\s0\fR website at \fIhttp://www.fs.net/\fR.
.SH "BUGS"
.IX Header "BUGS"
\&\fBvidb\fR should really recreate any publicly-readable versions
of user authentication databases (either by parsing
\&\fIsfsauthd_config\fR for \fB\-pub=...\fR options to
\&\fBUserfile\fR directives or signaling \fBsfsauthd\fR).
Currently you must manually kill \fBsfssd\fR or \fBsfsauthd\fR
for this to happen.
.PP
While \fBvidb\fR attempts to make the smallest number of changes
to a database, editing sessions that add or remove a large number of
records can potentially exhaust resources such as locks.  Sites with
large user databases can tune the database by creating a file called
\&\fI\s-1DB_CONFIG\s0\fR in the database directory.
The specifics of the configuration file are documented in the
Sleepycat database documentation.  As an example, if performance is
slow and you run out of locks, you can set the cache size to 128MB and
increase the number of locks with the following \fI\s-1DB_CONFIG\s0\fR file:
.PP
.Vb 3
\&  set_cachesize 0 134217728 1
\&  set_lk_max_locks 50000
\&  set_lk_max_objects 50000
.Ve
.PP
When editing a database, \fBvidb\fR creates a temporary text file
in the \fI/tmp\fR directory.  For huge databases, it is conceivable
that \fI/tmp\fR does not have enough space.  If this happens,
\&\fI/tmp\fR can be overridden with the \fB\s-1TMPDIR\s0\fR environment
variable.
.SH "AUTHOR"
.IX Header "AUTHOR"
sfsdev@redlab.lcs.mit.edu


sfsauthd/vidb.C
/* $Id: vidb.C,v 1.35 2004/06/17 21:10:24 dm Exp $ */

/*
 *
 * Copyright (C) 2001-2003 David Mazieres (dm@uun.org)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 *
 */

#include "authdb.h"
#include "sfsmisc.h"
#include "sfskeymisc.h"
#include "init.h"
#include "rxx.h"

static str editor;
static str tmppath;
static str db;
static ptr<authdb> targetdb;
static ptr<audblock> dblock;
static struct stat tmpsb;
static bool no_delete;
static ptr<authcursor> targetac;
static int killed;

static bool opt_runrecov;
static bool opt_wait;
static bool opt_salvage;
static bool opt_add;
static str opt_file;

EXITFN (cleanup);
static void
cleanup ()
{
  if (tmppath && !no_delete) {
    unlink (tmppath);
    tmppath = tmppath << "~";
    unlink (tmppath);
  }
  dblock = NULL;
  targetac = NULL;
  targetdb = NULL;
}

static void
setkilled (int sig)
{
  killed = 1;
}

static void
sethandlers ()
{
  signal (SIGINT, setkilled);
  signal (SIGTERM, setkilled);
}

static int
tmpinit ()
{
  const char *tmpdir = getenv ("TMPDIR");
  if (!tmpdir)
    tmpdir = "/tmp";
  char *pathbuf;
  {
    str tmpl (strbuf ("%s/%sXXXXXXXXXX", tmpdir, progname.cstr ()));
    pathbuf = xstrdup (tmpl);
  } 
  mode_t m = umask (077);
  int fd = mkstemp (pathbuf);
  if (fd < 0)
    fatal ("%s: %m\n", pathbuf);
  umask (m);
  tmppath = pathbuf;
  xfree (pathbuf);
  return fd;
}

static void
setdb ()
{
  str file = db;
  if (!file || !file.len ())
    fatal ("missing/invalid database name\n");
  if (!strchr (file, '/'))
    file = strbuf ("./") << db;
  targetdb = authdb_alloc (file);
  if (!targetdb)
    fatal << db << ": file extension not supported type\n";
}

static void
memberquery (str user)
{
  setdb ();
  vec<str> groups;
  
  ptr<authcursor> ac = targetdb->open (0, 0600, dblock);
  ac->find_groups_member (&groups, user);
  for (str *sp = groups.base (); sp < groups.lim (); sp++)
    printf ("%s\n", sp->cstr ());
}

static bool
lockit ()
{
  setdb ();
  if (db[db.len () - 1] == '/' && !opt_salvage
      && access (db, 0) < 0 && errno == ENOENT)
    mkdir (db, 0700);

  dblock = targetdb->lock (false);
  if (!dblock) {
    if (!opt_wait) {
      warn << db << ": could not lock\n";
      return false;
    }
    warn << ": waiting for lock on " << db << "...";
    if (dblock = targetdb->lock (opt_wait))
      warnx << " acquired\n";
  }

  return dblock;
}

static void
copyit (int fd)
{
  u_int32_t flags;
  if (opt_salvage)
    flags = authdb::AUDB_NORECOV;
  else {
    flags = authdb::AUDB_CREATE;
    if (opt_runrecov)
      flags |= authdb::AUDB_RUNRECOV;
  }

  ptr<authcursor> ac = targetdb->open (flags, 0600, dblock);
  if (!ac)
    fatal << db << ": " << strerror (errno) << "\n";

  int n;
  if ((n = fcntl (fd, F_GETFL)) >= 0)
    fcntl (fd, F_SETFL, n & ~O_NONBLOCK);

  bool pe = false;
  FILE *out = fdopen (fd, "w");
  while (ac->next (&pe)) {
    str line = authdbrec2str (&ac->ae);
    assert (line);
    if (fprintf (out, "%s\n", line.cstr ()) < 0) {
      if (tmppath)
	fatal << tmppath << ": " << strerror (errno) << "\n";
      else
	fatal << "stdout: " << strerror (errno) << "\n";
    }
  }
  if (pe)
    fatal << db << ": error reading/parsing file\n";

  fflush (out);
  if (!opt_salvage && fstat (fd, &tmpsb) < 0)
    fatal << tmppath << ": " << strerror (errno) << "\n";
  fclose (out);
}

static void
checklock ()
{
  if (!dblock->ok ()) {
    no_delete = true;
    fatal << "lock was stolen -- file saved in " << tmppath << "\n";
  }
}

static void
addrecords (str file)
{
  bool chatter = isatty (errfd);
  off_t filesize = 0;

  ptr<authcursor_file> vic;
  if (file) {
    if (access (file, R_OK) < 0)
      fatal ("%s: %m\n", file.cstr ());
    vic = New refcounted<authcursor_file> (file, 0, ptr<audblock> (NULL));
  }
  else {
    file = "(stdin)";
    vic = New refcounted<authcursor_file> (file, 0, ptr<audblock> (NULL), 0);
  }
  vic->linemax = 0;

  struct stat sb;
  if (!chatter || fstat (vic->fd, &sb) < 0 || !S_ISREG (sb.st_mode))
    chatter = false;
  else
    filesize = sb.st_size;

  targetac = targetdb->open ((authdb::AUDB_WRITE
			      | authdb::AUDB_CREATE
			      | authdb::AUDB_ASYNC),
			     0600, dblock);
  if (!targetac)
    fatal << db << ": could not open database for writing\n";

  sethandlers ();

  int pct = -1;
  if (chatter)
    warnx ("adding records...");
  while (vic->next () && !killed) {
    targetac->ae = vic->ae;
    if (!targetac->update ())
      fatal ("Database update failed\n");
    if (chatter) {
      int npct = vic->buf.byteno () / (filesize / 100);
      if (npct != pct) {
	pct = npct;
	warnx (" % 3d%%\b\b\b\b\b", pct);
      }
    }
  }
  if (chatter)
    warnx (killed ? " interrupted\n" : " done \n");
  if (killed)
    targetac = NULL;
  else if (!targetac->commit ())
    fatal ("Database update failed\n");

  exit (0);
}

static int
finish ()
{
  struct stat sb;
  if (stat (tmppath, &sb) < 0)
    fatal << tmppath << ": " << strerror (errno) << "\n";
  if (stat_unchanged (&sb, &tmpsb)) {
    if (isatty (1))
      warn << tmppath << " unchanged\n";
    exit (0);
  }

  no_delete = true;
    
  checklock ();

  ptr<authdb> vidb = New refcounted<authdb_file> (tmppath);
  ptr<authcursor> vic = vidb->open (0);
  if (!vic)
    fatal << tmppath << ": " << strerror (errno) << "\n";

  bool chatter = isatty (errfd);
  if (chatter)
    warnx ("copying...");

  targetac = targetdb->open ((authdb::AUDB_WRITE
			      | authdb::AUDB_CREATE
			      | authdb::AUDB_TRUNC),
			     0600, dblock);
  if (!targetac)
    fatal << db << ": could not open database for writing\n"
	  << "file saved in " << tmppath << "\n";

  int badline = 1;
  bool pe = false;
  bhash<str> keys;
  for (vic->reset (); vic->next (&pe) && !pe; badline++) {
    str aek = aekey (vic->ae);
    if (!keys.insert (aek)) {
      warn << tmppath << ":" << badline
	   << ": duplicate entry for " << aek << "\n";
      pe = true;
      break;
    }
    targetac->ae = vic->ae;
    if (!targetac->update ()) {
      targetac = NULL;
      fatal << "database update failed, file saved in " << tmppath << "\n";
    }
  }
  keys.clear ();

  if (chatter)
    warnx (" done\n");

  if (pe) {
    targetac = NULL;
    no_delete = false;
    return badline;
  }

  checklock ();

  if (chatter)
    warnx ("committing...");
  if (!targetac->commit (chatter ? badline : 0))
    fatal << "commit failed, file saved in " << tmppath << "\n";

  no_delete = false;
  targetac = NULL;
  targetdb = NULL;
  exit (0);
}

static void
usage ()
{
  warnx << "usage: " << progname
	<< " [-w] [-R] {-S | -a [-f file] | [-e editor]} dbfile\n";
  exit (1);
}

bool
run_editor (str file, int line)
{
  static rxx edplus ("^(vi|nvi|.*emacs.*|vim|ee|ex|mg|pico)$");

  if (strchr (editor, ' ')) {
    str cmd (strbuf () << editor << " " << file);
    return !system (cmd);
  }

  const char *av[] = { editor, file, NULL, NULL };
  str plusarg;
  if (line > 0 && edplus.match (editor)) {
    plusarg = strbuf ("+%d", line);
    av[1] = plusarg;
    av[2] = file;
  }

  make_sync (0);
  make_sync (1);
  pid_t pid = spawn (editor, av);
  int status;
  return pid != -1 && waitpid (pid, &status, 0) == pid && !status;
}

static bool
get_yesno (const str &prompt) 
{
  for (;;) {
    str r = getline (prompt, NULL);
    if (!r || r.len () == 0) 
      return false;
    const char *cp = r.cstr ();
    if (*cp == 'Y' || *cp == 'y')
      return true;
    if (*cp == 'N' || *cp == 'n')
      return false;
  }
}

int
main (int argc, char **argv)
{
  setprogname (argv[0]);
  sfsconst_init ();

  str opt_memberquery;
  editor = getenv ("EDITOR");
  int ch;
  while ((ch = getopt (argc, argv, "RSwg:m:e:f:a")) != -1)
    switch (ch) {
    case 'S':
      opt_salvage = true;
      break;
    case 'R':
      opt_runrecov = true;
      break;
    case 'w':
      opt_wait = true;
      break;
    case 'g':
      if (opt_memberquery)
	usage ();
      opt_memberquery = optarg;
      break;
    case 'f':
      opt_file = optarg;
      break;
    case 'e':
      editor = optarg;
      break;
    case 'a':
      opt_add = true;
      break;
    default:
      usage ();
    }

  if ((opt_salvage && opt_add) || (opt_file && !opt_add))
    usage ();

  if (optind + 1 != argc)
    usage ();
  db = argv[optind];

  if (!editor)
    editor = "vi";
  if (str edpath = find_program (editor))
    editor = edpath;
  else
    fatal << editor << ": " << strerror (errno) << "\n";

  if (opt_memberquery) {
    if (opt_wait || opt_add || opt_file || opt_salvage)
      usage ();
    memberquery (opt_memberquery);
    exit (0);
  }

  if (!lockit () && !opt_salvage)
    fatal << db << ": could not lock database\n";

  if (opt_salvage) {
    copyit (1);
    exit (0);
  }

  if (opt_add) {
    addrecords (opt_file);
    exit (0);
  }

  int fd = tmpinit ();
  copyit (fd);

  int line = 0;
  for (;;) {
    if (!run_editor (tmppath, line))
      fatal << editor << ": abnormal exit\n";
    line = finish ();
    warn << "Database update refused\n";
    if (!get_yesno ("Retry database edit [y/N]? ")) {
      warn << "Database update aborted.\n";
      exit (0);
    }
  }
}
